File: src/ScOutfitter.Core/Catalog.cs
using System.Globalization;
using System.Text.Json.Nodes;

namespace ScOutfitter.Core;

public enum Kind
{
    Gun,
    Shield,
    PowerPlant,
    Cooler,
    QuantumDrive,
    MissileRack,
    Missile,
    Radar,
}

public static class KindNames
{
    public static string Label(this Kind kind) => kind switch
    {
        Kind.Gun => "gun",
        Kind.Shield => "shield",
        Kind.PowerPlant => "power plant",
        Kind.Cooler => "cooler",
        Kind.QuantumDrive => "quantum drive",
        Kind.MissileRack => "missile rack",
        Kind.Missile => "missile",
        Kind.Radar => "radar",
        _ => kind.ToString(),
    };
}

public sealed record Offer(int TerminalId, string TerminalName, int Price);

/// <summary>One purchasable (or stock) part with the stats the optimizer scores on.</summary>
public sealed class Component
{
    public required string Name { get; init; }
    public required Kind Kind { get; init; }
    public required int Size { get; init; }
    public string Grade { get; init; } = "?";
    public string Class { get; init; } = string.Empty;
    public Dictionary<string, double> Stats { get; } = new(StringComparer.Ordinal);
    public bool Ammo { get; init; }
    public string Signal { get; init; } = string.Empty;
    public double PowerDraw { get; init; }
    public double CoolantDraw { get; init; }
    public List<Offer> Offers { get; } = [];
    public HashSet<string> Tags { get; init; } = new(StringComparer.Ordinal);
    public HashSet<string> RequiredTags { get; init; } = new(StringComparer.Ordinal);

    public bool Buyable => Offers.Count > 0;

    public int? Cheapest => Offers.Count == 0 ? null : Offers.Min(o => o.Price);

    public double Stat(string key) => Stats.TryGetValue(key, out double v) ? v : 0;

    public string Summary()
    {
        CultureInfo inv = CultureInfo.InvariantCulture;
        return Kind switch
        {
            Kind.Gun => $"{Stat("dps"):0} dps, {Stat("range"):0} m{(Ammo ? ", ammo" : "")}",
            Kind.Shield => $"{Stat("hp"):0} hp, {Stat("regen"):0}/s regen",
            Kind.PowerPlant => $"{Stat("power"):0} power segments",
            Kind.Cooler => $"{Stat("coolant"):0} coolant segments",
            Kind.QuantumDrive => string.Format(inv, "{0:0} Mm/s, spool {1:0.0}s, {2:0.00} fuel/Gm",
                Stat("speed") / 1e6, Stat("spool"), Stat("fuel_rate") * 1e9),
            Kind.MissileRack => $"{Stat("count"):0}x S{Stat("missile_size"):0}",
            Kind.Missile => $"{Stat("damage"):0} dmg, {Stat("speed"):0} m/s, {Stat("range") / 1000:0} km, {Signal}",
            Kind.Radar => $"aim assist {Stat("aim_range"):0} m, EM {Stat("em"):0}",
            _ => string.Empty,
        };
    }

    public override string ToString() => $"{Name} (S{Size} {Kind})";
}

/// <summary>Turns wiki items into <see cref="Component"/> records grouped by kind.</summary>
public static class Catalog
{
    public static readonly (string WikiType, Kind Kind)[] Types =
    [
        ("QuantumDrive", Kind.QuantumDrive),
        ("Shield", Kind.Shield),
        ("PowerPlant", Kind.PowerPlant),
        ("Cooler", Kind.Cooler),
        ("WeaponGun", Kind.Gun),
        ("MissileLauncher", Kind.MissileRack),
        ("Missile", Kind.Missile),
        ("Radar", Kind.Radar),
    ];

    public static async Task<Dictionary<Kind, List<Component>>> LoadAsync(
        WikiClient client, IProgress<string>? progress = null, CancellationToken ct = default)
    {
        var catalog = new Dictionary<Kind, List<Component>>();
        foreach ((string wikiType, Kind kind) in Types)
        {
            progress?.Report($"Loading {kind.Label()}s ...");
            List<JsonNode> items = await client.ItemsAsync(wikiType, null, ct).ConfigureAwait(false);
            catalog[kind] = Build(kind, items);
        }

        return catalog;
    }

    /// <summary>
    /// Stock parts (e.g. Regulus) are flagged as variants on the wiki and the same name can appear
    /// several times (loot/paint variants) with the shop offers on only one of them, so entries are
    /// merged by (name, size) instead of filtering on is_base_variant.
    /// </summary>
    public static List<Component> Build(Kind kind, IEnumerable<JsonNode> items)
    {
        var byKey = new Dictionary<(string, int), Component>();
        foreach (JsonNode item in items)
        {
            Component? comp = FromJson(kind, item);
            if (comp is null)
            {
                continue;
            }

            (string Name, int Size) key = (comp.Name, comp.Size);
            if (byKey.TryGetValue(key, out Component? existing))
            {
                existing.Offers.AddRange(comp.Offers);
            }
            else
            {
                byKey[key] = comp;
            }
        }

        return byKey.Values.ToList();
    }

    public static Component? FromJson(Kind kind, JsonNode item)
    {
        var stats = new Dictionary<string, double>(StringComparer.Ordinal);
        bool ammo = false;
        string signal = string.Empty;

        switch (kind)
        {
            case Kind.Gun:
            {
                JsonNode? w = item.At("vehicle_weapon");
                double dps = w.Arr("modes").Select(m => m.Num("damage_per_second")).DefaultIfEmpty(0).Max();
                if (dps <= 0)
                {
                    return null;
                }

                stats["dps"] = dps;
                stats["range"] = w.Num("range");
                ammo = w.Num("capacity") > 0;
                break;
            }

            case Kind.Shield:
            {
                double hp = item.Num("shield", "max_health");
                if (hp <= 0)
                {
                    return null;
                }

                stats["hp"] = hp;
                stats["regen"] = item.Num("shield", "regen_rate");
                break;
            }

            case Kind.PowerPlant:
            {
                double gen = item.Num("power_plant", "power_segment_generation");
                if (gen <= 0)
                {
                    return null;
                }

                stats["power"] = gen;
                break;
            }

            case Kind.Cooler:
            {
                double gen = item.Num("cooler", "coolant_segment_generation");
                if (gen <= 0)
                {
                    return null;
                }

                stats["coolant"] = gen;
                break;
            }

            case Kind.QuantumDrive:
            {
                JsonNode? q = item.At("quantum_drive");
                JsonNode? j = q.At("standard_jump");
                double speed = j.Num("drive_speed");
                if (speed <= 0)
                {
                    return null;
                }

                stats["speed"] = speed;
                stats["spool"] = j.Num("spool_up_time");
                stats["cooldown"] = j.Num("cooldown_time");
                double accel = j.Num("stage_two_accel_rate");
                stats["accel"] = accel > 0 ? accel : j.Num("stage_one_accel_rate");
                stats["fuel_rate"] = q.Num("fuel_rate"); // fuel units per metre
                break;
            }

            case Kind.MissileRack:
            {
                double count = item.Num("missile_rack", "missile_count");
                if (count <= 0)
                {
                    return null;
                }

                // do not filter on required_tags: the MSD-322 entry that carries the shop offers has them
                stats["count"] = count;
                stats["missile_size"] = item.Num("missile_rack", "missile_size");
                break;
            }

            case Kind.Missile:
            {
                JsonNode? m = item.At("missile");
                double dmg = m.Num("damage_total");
                if (dmg <= 0)
                {
                    return null;
                }

                double speed = m.Num("flight", "speed");
                if (speed <= 0)
                {
                    speed = m.Num("speed");
                }

                stats["damage"] = dmg;
                stats["speed"] = speed;
                stats["range"] = m.Num("flight", "range");
                stats["lock_time"] = m.Num("lock_time");
                signal = m.Str("signal_type");
                if (signal.Length == 0)
                {
                    signal = "?";
                }

                break;
            }

            case Kind.Radar:
            {
                JsonNode? r = item.At("radar");
                double aim = r.Num("aim_assist", "distance_max_assignment");
                double[] sens = new[] { "infrared", "cross_section", "electromagnetic" }
                    .Select(k => r.Num("sensitivity", k)).Where(v => v > 0).ToArray();
                if (aim <= 0 && sens.Length == 0)
                {
                    return null;
                }

                // sensitivity is identical across all radars in 4.x; what differs is the aim-assist
                // range (the number Erkul shows in green) and the EM signature
                stats["aim_range"] = aim;
                stats["sensitivity"] = sens.Length > 0 ? sens.Average() : 0;
                stats["em"] = item.Num("emission", "em_max");
                break;
            }

            default:
                return null;
        }

        var comp = new Component
        {
            Name = item.Str("name"),
            Kind = kind,
            Size = item.Int("size"),
            Grade = item.StrOrNull("grade") is { Length: > 0 } g ? g : "?",
            Class = item.Str("class"),
            Ammo = ammo,
            Signal = signal,
            PowerDraw = item.Num("resource_network", "usage", "power", "max"),
            CoolantDraw = item.Num("resource_network", "usage", "coolant", "max"),
            Tags = item.StrSet("tags"),
            RequiredTags = item.StrSet("required_tags"),
        };
        foreach ((string k, double v) in stats)
        {
            comp.Stats[k] = v;
        }

        foreach (JsonNode o in item.Arr("uex_prices", "purchase"))
        {
            int price = o.Int("price_buy");
            if (price > 0)
            {
                comp.Offers.Add(new Offer(o.Int("terminal_id"), o.Str("terminal_name"), price));
            }
        }

        return comp.Name.Length == 0 ? null : comp;
    }
}

File: outfitter/catalog.py
"""Normalize wiki items into flat Component records with a shop offer list."""
from __future__ import annotations

from dataclasses import dataclass, field

from . import api

# wiki item type -> our slot kind
TYPES = {
    "QuantumDrive": "quantum_drive",
    "Shield": "shield",
    "PowerPlant": "power_plant",
    "Cooler": "cooler",
    "WeaponGun": "gun",
    "MissileLauncher": "missile_rack",
    "Missile": "missile",
    "Radar": "radar",
}
KINDS = list(TYPES.values())


@dataclass
class Offer:
    terminal_id: int
    terminal_name: str
    price: int


@dataclass
class Component:
    name: str
    kind: str
    size: int
    grade: str
    cls: str
    stats: dict = field(default_factory=dict)
    power_draw: float = 0.0
    coolant_draw: float = 0.0
    offers: list[Offer] = field(default_factory=list)
    tags: frozenset = frozenset()           # e.g. {"LaserRepeater", "Wolf_Gun"}
    required_tags: frozenset = frozenset()  # bespoke parts: only fit ports that carry these

    @property
    def buyable(self) -> bool:
        return bool(self.offers)

    @property
    def cheapest(self) -> int | None:
        return min((o.price for o in self.offers), default=None)

    def summary(self) -> str:
        s = self.stats
        if self.kind == "gun":
            return f"{s['dps']:.0f} dps, {s['range']:.0f} m{', ammo' if s['ammo'] else ''}"
        if self.kind == "shield":
            return f"{s['hp']:.0f} hp, {s['regen']:.0f}/s regen"
        if self.kind == "power_plant":
            return f"{s['power']:.0f} power segments"
        if self.kind == "cooler":
            return f"{s['coolant']:.0f} coolant segments"
        if self.kind == "quantum_drive":
            return (f"{s['speed'] / 1e6:.0f} Mm/s, spool {s['spool']:.1f}s, "
                    f"{s['fuel_rate'] * 1e9:.2f} fuel/Gm")
        if self.kind == "missile_rack":
            return f"{s['count']}x S{s['missile_size']}"
        if self.kind == "missile":
            return f"{s['damage']:.0f} dmg, {s['speed']:.0f} m/s, {s['range'] / 1000:.0f} km, {s['signal']}"
        if self.kind == "radar":
            return f"sensitivity {s['sensitivity']:.2f}"
        return ""


def _usage(item: dict) -> tuple[float, float]:
    rn = item.get("resource_network") or {}
    u = rn.get("usage") or {}
    p = (u.get("power") or {}).get("max") or 0.0
    c = (u.get("coolant") or {}).get("max") or 0.0
    return float(p), float(c)


def _stats(kind: str, item: dict) -> dict | None:
    if kind == "gun":
        w = item.get("vehicle_weapon") or {}
        modes = w.get("modes") or []
        dps = max((m.get("damage_per_second") or 0) for m in modes) if modes else 0
        if dps <= 0:
            return None
        return {"dps": float(dps), "range": float(w.get("range") or 0),
                "ammo": bool(w.get("capacity")), "weapon_type": w.get("type") or ""}
    if kind == "shield":
        s = item.get("shield") or {}
        if not s.get("max_health"):
            return None
        return {"hp": float(s["max_health"]), "regen": float(s.get("regen_rate") or 0)}
    if kind == "power_plant":
        p = item.get("power_plant") or {}
        gen = p.get("power_segment_generation") or 0
        return {"power": float(gen)} if gen else None
    if kind == "cooler":
        c = item.get("cooler") or {}
        gen = c.get("coolant_segment_generation") or 0
        return {"coolant": float(gen)} if gen else None
    if kind == "quantum_drive":
        q = item.get("quantum_drive") or {}
        j = q.get("standard_jump") or {}
        if not j.get("drive_speed"):
            return None
        return {"speed": float(j["drive_speed"]), "spool": float(j.get("spool_up_time") or 0),
                "cooldown": float(j.get("cooldown_time") or 0),
                "accel": float(j.get("stage_two_accel_rate") or j.get("stage_one_accel_rate") or 0),
                "fuel_rate": float(q.get("fuel_rate") or 0)}  # fuel units per metre
    if kind == "missile_rack":
        r = item.get("missile_rack") or {}
        if not r.get("missile_count"):
            return None
        # do not filter on required_tags: the MSD-322 entry that carries the shop offers has them
        return {"count": int(r["missile_count"]), "missile_size": int(r["missile_size"])}
    if kind == "missile":
        m = item.get("missile") or {}
        dmg = m.get("damage_total") or 0
        if not dmg or item.get("sub_type") == "Torpedo" and not m.get("speed"):
            return None
        fl = m.get("flight") or {}
        return {"damage": float(dmg), "speed": float(fl.get("speed") or m.get("speed") or 0),
                "range": float(fl.get("range") or 0), "signal": m.get("signal_type") or "?",
                "lock_time": float(m.get("lock_time") or 0)}
    if kind == "radar":
        r = item.get("radar") or {}
        sens = r.get("sensitivity") or {}
        vals = [sens.get(k) for k in ("infrared", "cross_section", "electromagnetic") if sens.get(k)]
        if not vals:
            return None
        return {"sensitivity": sum(vals) / len(vals), "sub_type": item.get("sub_type") or ""}
    return None


def _offers(item: dict) -> list[Offer]:
    out = []
    for o in ((item.get("uex_prices") or {}).get("purchase") or []):
        if o.get("price_buy"):
            out.append(Offer(int(o["terminal_id"]), o.get("terminal_name") or "", int(o["price_buy"])))
    return out


def load_catalog(kinds: set[str] | None = None) -> dict[str, list[Component]]:
    """{kind: [Component]} for every kind in TYPES (or the requested subset)."""
    catalog: dict[str, list[Component]] = {}
    for wiki_type, kind in TYPES.items():
        if kinds and kind not in kinds:
            continue
        by_key: dict[tuple[str, int], Component] = {}
        for item in api.wiki_items(wiki_type):
            stats = _stats(kind, item)
            if stats is None:
                continue
            # stock parts (e.g. Regulus) are flagged as variants on the wiki and the same name can
            # appear several times (loot/paint variants) with the shop offers on only one of them,
            # so merge by (name, size) instead of filtering on is_base_variant
            power, coolant = _usage(item)
            comp = Component(item["name"], kind, int(item.get("size") or 0),
                             str(item.get("grade") or "?"), str(item.get("class") or ""),
                             stats, power, coolant, _offers(item),
                             frozenset(item.get("tags") or []), frozenset(item.get("required_tags") or []))
            key = (comp.name, comp.size)
            if key in by_key:
                by_key[key].offers.extend(comp.offers)
            else:
                by_key[key] = comp
        catalog[kind] = list(by_key.values())
    return catalog
